pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/SegmentIndexCreationDriverImpl.java

```diff
@@ -133,6 +135,30 @@ public void build() throws Exception {
     }
   }
 
+  private void augmentSchemaWithDerivedColumns() {
+    HllConfig hllConfig = config.getHllConfig();
+    if (hllConfig != null) {
+      for (Entry<String, String> entry : hllConfig.getHllDeriveFieldNameMap().entrySet()) {
+        String originFieldName = entry.getKey();
+        String derivedFieldName = entry.getValue();
+        dataSchema.addField(derivedFieldName,
+                new HllDerivedFieldSpec(derivedFieldName, hllConfig.getHllLog2m()));
+      }
+    }
+  }
+
+  private void populateDefaultDerivedColumnValues(GenericRow row) throws IOException {
+    HllConfig hllConfig = config.getHllConfig();
+    if (hllConfig != null) {
+      for (Entry<String, String> entry : hllConfig.getHllDeriveFieldNameMap().entrySet()) {
+        String originFieldName = entry.getKey();
+        String derivedFieldName = entry.getValue();
+        byte[] derivedFieldValue = HllUtil.singleValueHllInBytes(hllConfig.getHllLog2m(), row.getValue(originFieldName));
+        row.addField(derivedFieldName, derivedFieldValue);
+      }
+    }
+  }
+
   public void buildStarTree() throws Exception {
```

Comment:
For non-star tree based segment generation, log a warn message (only once) if the hll configuration is specified. Something along the lines - "HLL configuration is works only with star tree. Star Tree generation is not specified. This will not add derived hll column"